protocol/x/clob/e2e/withdrawal_gating_test.go

```diff
@@ -0,0 +1,231 @@
+package clob_test
+
+import (
+	"testing"
+
+	abcitypes "github.com/cometbft/cometbft/abci/types"
+	sdktypes "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/dydxprotocol/v4-chain/protocol/daemons/liquidation/api"
+
+	"github.com/cometbft/cometbft/types"
+
+	testapp "github.com/dydxprotocol/v4-chain/protocol/testutil/app"
+	clobtest "github.com/dydxprotocol/v4-chain/protocol/testutil/clob"
+	"github.com/dydxprotocol/v4-chain/protocol/testutil/constants"
+	assettypes "github.com/dydxprotocol/v4-chain/protocol/x/assets/types"
+	clobtypes "github.com/dydxprotocol/v4-chain/protocol/x/clob/types"
+	feetiertypes "github.com/dydxprotocol/v4-chain/protocol/x/feetiers/types"
+	perptypes "github.com/dydxprotocol/v4-chain/protocol/x/perpetuals/types"
+	prices "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+	sendingtypes "github.com/dydxprotocol/v4-chain/protocol/x/sending/types"
+	satypes "github.com/dydxprotocol/v4-chain/protocol/x/subaccounts/types"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWithdrawalGating(t *testing.T) {
+	tests := map[string]struct {
+		// State.
+		subaccounts                   []satypes.Subaccount
+		marketIdToOraclePriceOverride map[uint32]uint64
+
+		// Parameters.
+		placedMatchableOrders     []clobtypes.MatchableOrder
+		liquidatableSubaccountIds []satypes.SubaccountId
+		negativeTncSubaccountIds  []satypes.SubaccountId
+
+		// Configuration.
+		liquidationConfig clobtypes.LiquidationsConfig
+		liquidityTiers    []perptypes.LiquidityTier
+		perpetuals        []perptypes.Perpetual
+		clobPairs         []clobtypes.ClobPair
+
+		// Expectations.
+		expectedSubaccounts                      []satypes.Subaccount
+		expectedWithdrawalsGated                 bool
+		expectedNegativeTncSubaccountSeenAtBlock uint32
+		expectedErr                              string
+	}{
+		`Can place a liquidation order that is unfilled and cannot be deleveraged due to
+		non-overlapping bankruptcy prices, withdrawals are gated`: {
```

Comment:
with the current implementation, liquidation order doesn't need to be unfilled and deleverage can still happen for withdrawals to be gated, right? since this is happening in step 8 which is after liquidations/deleveraging steps. 

can we add tests for these?

Reply:
yeah I'll add tests for all variants of the below possible scenarios that can happen for verifying withdrawals are gated / not gated:
- liquidation order unfilled, partial deleveraging
- liquidation order filled, deleveraging fails
- liquidation order filled, partial deleveraging